File: kalitech-engine/src/main/resources/kalitech/engine/bootstrap/Meta.js
// FILE: resources/kalitech/engine/bootstrap/Meta.js
"use strict";

const {isObj} = require("./Util.js");
const {tryReadKaliModFromFsNearModule, tryReadKaliModFromAssets} = require("./Assets.js");

function normalizeMeta(exp, fallbackName, moduleId, engine) {
    const fromFs = moduleId ? tryReadKaliModFromFsNearModule(moduleId) : null;
    const fromAssets = (!fromFs && engine && moduleId) ? tryReadKaliModFromAssets(engine, moduleId) : null;
    const fromExport = (exp && exp.META && isObj(exp.META)) ? exp.META : null;

    const src = fromFs || fromAssets || fromExport;

    const name = (src && src.name) ? String(src.name) : String(fallbackName || "");
    const globalName = (src && src.globalName) ? String(src.globalName) : "";
    const version = (src && src.version) ? String(src.version) : "0.0.0";
    const description = (src && src.description) ? String(src.description) : "";
    const engineMin = (src && src.engineMin) ? String(src.engineMin) : "";
    return {name, globalName, version, description, engineMin};
}

module.exports = {normalizeMeta};

File: kalitech-engine/src/main/resources/kalitech/engine/bootstrap/Meta.lua
local M = {}

local function normalizeMeta(_, exported, fallbackName, moduleId)
    local source = exported and exported.META
    if source ~= nil and KTypeOf(source) ~= "table" then
        error("[bootstrap] module META must be a Lua table: " .. tostring(moduleId), 0)
    end
    source = source or {}

    return {
        moduleId = tostring(source.moduleId or moduleId or fallbackName or ""),
        name = tostring(source.name or fallbackName or moduleId or ""),
        version = tostring(source.version or "0.0.0"),
        description = tostring(source.description or ""),
        engineMin = tostring(source.engineMin or "")
    }
end

M.normalizeMeta = normalizeMeta
return M
